### PR TITLE
Tables migration guide

### PR DESCRIPTION
**Pull request by seankane-msft**

### CODE REVIEW

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    "name1": "myTableName1",
+    "name2": "myTableName2",
+}
+
+for table in service_client.query_tables(query_filter, parameters=parameters):
+    print(table.name)
+```
+
+### Entity Level Scenarios
+
```
Comment (mccoyp) [inline sdk/tables/azure-data-tables/migration_guide.md:137]: Is there anything to put here?

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    "BooleanProperty": False,
+    "IntegerProperty": 31,
+    "FloatProperty": 3.14159,
+    "BinaryProperty": b"binary",
+    "GuidProperty": uuid.uuid4(),
+    "DatetimeProperty": datetime.datetime.now(),
+}
+
+endpoint = "https://{}.table.cosmos.azure.com".format(account_name)
+# endpoint = "https://{}.table.core.windows.net".format(account_name)
```
Comment (mccoyp) [inline sdk/tables/azure-data-tables/migration_guide.md:181]: Is this comment meant to be in the sample?

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+#### Create and Delete
+
+In `azure-cosmosdb-table`:
+```python
+from azure.cosmosdb.table import TableService
+client = TableService(...)
+table_name = "tableName"
+created = client.create_table(table_name)
+# create_table returns True if a new table was created, False if the table already exists
```
Comment (swathipil) [inline sdk/tables/azure-data-tables/migration_guide.md:81]: nit, move comment above call for consistency with rest of comments.

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+# Delete
+table_name = "deleteTableName"
+client.delete_table(table_name)
+# delete_table returns True if a table was deleted, False if the table does not exist
```
Comment (swathipil) [inline sdk/tables/azure-data-tables/migration_guide.md:86]: move comment above call for consistency

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+```python
+from azure.data.tables import TableServiceClient, TableClient
+
+# You can create a table from the TableServiceClient or TableClient
+service_client = TableServiceClient.from_connection_string(conn_str)
+table_client = service_client.create_table("tableName")
+service_client.delete_table("tableName")
+
+# The create_table_if_not_exists
+table_client = TableClient.from_connection_string(conn_str, table_name="tableName")
```
Comment (swathipil) [inline sdk/tables/azure-data-tables/migration_guide.md:99]: for clarity, what is the difference b/w using both clients to create a table. Why can both be used to do this?

Maybe add that explanation at the bottom or top of this section (### Create and delete)

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+credential = AzureNamedKeyCredential(key=key, name=name)
+
+table_client = TableClient(account_url, "tablename", credential=credential)
+
+table_service_client = TableServiceClient(account_url, credential=credential)
+```
+
+
+### Table Level Scenarios
```
Comment (swathipil) [inline sdk/tables/azure-data-tables/migration_guide.md:71]: in general, for scenarios below, answer the question:
Do Cosmos and Tables achieve the same thing? Then, why is it better to use Tables than Cosmos, esp from customer's standpoint?

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+table_client = service_client.create_table("tableName")
+service_client.delete_table("tableName")
+
+# The create_table_if_not_exists
+table_client = TableClient.from_connection_string(conn_str, table_name="tableName")
+table_client.create_table_if_not_exists()
+table_client.delete_table()
+```
+
+#### List and Query
```
Comment (swathipil) [inline sdk/tables/azure-data-tables/migration_guide.md:104]: A little text section on what has been updated from Cosmos, and WHY is Tables the better solution?

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+parameters = {
+    "name1": "myTableName1",
+    "name2": "myTableName2",
+}
+
+for table in service_client.query_tables(query_filter, parameters=parameters):
+    print(table.name)
+```
+
+### Entity Level Scenarios
```
Comment (swathipil) [inline sdk/tables/azure-data-tables/migration_guide.md:136]: Same thing here. Why should customers use Tables for each scenario? Better? Faster? Stronger?

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+    ("update", entity5, {"mode": "replace"}),
+]
+
+table_client.submit_transaction(operations)
+
+```
+
+## Additional samples
+
+More examples can be found at the [library home page](https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/tables/azure-data-tables/samples).
```
Comment (swathipil) [inline sdk/tables/azure-data-tables/migration_guide.md:289]:
```suggestion
More examples can be found at the [samples for azure-data-tables](https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/tables/azure-data-tables/samples).
```

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+# delete_table returns True if a table was deleted, False if the table does not exist
+```
+
+In `azure-data-tables`:
+```python
+from azure.data.tables import TableServiceClient, TableClient
+
+# You can create a table from the TableServiceClient or TableClient
+service_client = TableServiceClient.from_connection_string(conn_str)
+table_client = service_client.create_table("tableName")
```
[the remark this answers is not in the document]
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/migration_guide.md:95]: TableClient

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+service_client.delete_table("tableName")
+
+# The create_table_if_not_exists
+table_client = TableClient.from_connection_string(conn_str, table_name="tableName")
+table_client.create_table_if_not_exists()
+table_client.delete_table()
```
[the remark this answers is not in the document]
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/migration_guide.md:101]: No, the TableClient has a table name as an instantiation parameter, the TableServiceClient create_table method does take a table name parameter.

File: sdk/tables/azure-data-tables/migration_guide.md (hunk 1 above)
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/migration_guide.md:137]: Not really, these are just logical separations of the types of interactions with the service

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 12]
```diff
@@ -0,0 +1,289 @@
+# Guide for migrating to `azure-data-tables` from `azure-cosmosdb-table`
+
+This guide is intended to assist in the migration to `azure-data-tables` from `azure-cosmosdb-table`. It will focus on side-by-side comparisons for similar operations between the two packages.
+
+We assume that you are familiar with `azure-cosmosdb-table`. If not, please refer to the README for `azure-cosmosdb-table` rather than this guide.
+
+## Table of contents
+
```
[the remark this answers is not in the document]
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/migration_guide.md:8]: I have added this, thanks for pointing it out

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+The modern `azure-data-tables` client library also provides the ability to share in some of the cross-service improvements made to the Azure development experience, such as
+<!-- # Not used in this library, should we still include?
+- Using the new `azure-identity` library to share a single authentication approach between clients -->
+- A unified logging and diagnostics pipeline offering a common view of the activities across each of the client libraries
+
+### Performance improvements
+
+Use this section to advertise the performance improvements in new package when compared to the old one. Skip this section if no perf improvements are found yet.
+
```
[the remark this answers is not in the document]
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/migration_guide.md:27]: Anna will probably have this information, but most likely it will be removed.

File: sdk/tables/azure-data-tables/migration_guide.md (hunk 6 above)
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/migration_guide.md:71]: In general Cosmos and tables achieve the same thing, there are certain properties a user can set on Tables only (which we don't display here). It's mostly a customer choice.

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 14]
```diff
@@ -0,0 +1,298 @@
+# Guide for migrating to `azure-data-tables` from `azure-cosmosdb-table`
+
+This guide is intended to assist in the migration to `azure-data-tables` from `azure-cosmosdb-table`. It will focus on side-by-side comparisons for similar operations between the two packages.
+
+We assume that you are familiar with `azure-cosmosdb-table`. If not, please refer to the README for [`azure-cosmosdb-table`](https://github.com/Azure/azure-cosmos-table-python/tree/master/azure-cosmosdb-table) rather than this guide.
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/migration_guide.md:5]: Why would we be pointing them to the old repo? Wouldn't we be better of specifying the target audience as being people who have code built around `azure-cosmosdb-table` - and for those that do not, simply direct them to `azure-data-tables`?

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+
+To improve the development experience across Azure services, a set of uniform [design guidelines](https://azure.github.io/azure-sdk/general_introduction.html) were created for all languages to drive a consistent experience with established API patterns for all services. A set of [Python guidelines](https://azure.github.io/azure-sdk/python_design.html) were also introduced to ensure that Python clients have a natural and idiomatic feel with respect to the Python ecosystem. The new `azure-data-tables` follows these guidelines.
+
+### Cross Service SDK improvements
+
+The modern `azure-data-tables` client library also provides the ability to share in some of the cross-service improvements made to the Azure development experience, such as:
+- A unified logging and diagnostics pipeline offering a common view of the activities across each of the client libraries
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/migration_guide.md:29]: Maybe add here a unified authentication experience?

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+
+### Performance improvements
+
+Use this section to advertise the performance improvements in new package when compared to the old one. Skip this section if no perf improvements are found yet.
+@annatisch, should we remove this section?
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/migration_guide.md:34]: Yes remove this section for now.

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+
+### New features
+
+We have a variety of new features available in the new Azure Data Tables library:
+* Ability to submit a batch of operations with the `TableClient.submit_transaction` method
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/migration_guide.md:39]: This isn't a new feature though is it? T1 SDK could do this.

Comment (annatisch) [reply sdk/tables/azure-data-tables/migration_guide.md:39]: Maybe could rephrase it to something like "list-based transactions"

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+* Separates Table and Entity level operations into two clients, `TableServiceClient` and `TableClient`
+* Adds the ability to query tables and list entities explicitly.
+
+## Important changes
+
+### Package names and namespaces
+
+The package name has been changed from `azure-cosmosdb-table` to `azure-data-tables`. This package can be used to access and manipulate data stored in a CosmosDB Table account or an Azure Storage Tables account.
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/migration_guide.md:47]: "...stored in either a..."

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 19]
```diff
@@ -0,0 +1,298 @@
...
+
+### Client hierarchy and constructors
+
+In the interest of simplicity, there are only two clients, `TableServiceClient` for account-level interactions and `TableClient` for table-level interactions. This is in contrast to the single `ServiceClient` used for interacting with the account level operations and entity level operations. The two clients will simplify the method calls by not requiring a Table name to be included on each service call.
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/migration_guide.md:51]: I feel this line needs rephrasing. It's a bit contradictory to say "in the interest of simplicity there are only two clients" when that is in fact an increase in complexity by a factor of 2 ;)
How about something more like, "In order to provide an improved conceptual experience, we have divided the single `ServiceClient` into two clients; `TableServiceClient` and `TableClient`. The former provides access to all account-level operations..." etc etc

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 20]
```diff
@@ -0,0 +1,298 @@
...
+
+key = os.environ["TABLES_ACCOUNT_KEY"]
+name = os.environ["TABLES_ACCOUNT_NAME"]
+url = os.environ["TABLES_ACCOUNT_URL"]
+
+credential = AzureNamedKeyCredential(key=key, name=name)
+
+table_client = TableClient(account_url, "tablename", credential=credential)
+
+table_service_client = TableServiceClient(account_url, credential=credential)
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/migration_guide.md:77]: I would swap the `table_client` and `table_service_client` lines around.

File: sdk/tables/azure-data-tables/migration_guide.md (hunk 6 above)
Comment (annatisch) [reply sdk/tables/azure-data-tables/migration_guide.md:81]: @swathipil - are you asking why I would use CosmosTables vs StorageTables? Or why I would use T1 SDK vs T2 SDK?

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 21]
```diff
@@ -0,0 +1,298 @@
...
+from azure.data.tables import TableServiceClient, TableClient
+
+# You can create a table from the TableServiceClient or TableClient
+service_client = TableServiceClient.from_connection_string(conn_str)
+table_client = service_client.create_table("tableName")
+service_client.delete_table("tableName")
+
+# The create_table_if_not_exists is an alternative to create_table
+table_client = TableClient.from_connection_string(conn_str, table_name="tableName")
+table_client.create_table_if_not_exists()
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/migration_guide.md:110]: This is incorrect. `create_table_if_not_exists` is only on the `TableServiceClient`

File: sdk/tables/azure-data-tables/migration_guide.md (hunk 7 above)
Comment (annatisch) [reply sdk/tables/azure-data-tables/migration_guide.md:114]: Just to clarify, Cosmos _is_ tables. I guess you mean T1 vs T2?

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 22]
```diff
@@ -0,0 +1,298 @@
...
+    print(table.name)
+
+query_filter = "TableName eq @name1 or TableName gt @name2'
+parameters = {
+    "name1": "myTableName1",
+    "name2": "myTableName2",
+}
+
+for table in service_client.query_tables(query_filter, parameters=parameters):
+    print(table.name)
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/migration_guide.md:143]: You didn't demonstrate query in T1?

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 23]
```diff
@@ -0,0 +1,298 @@
...
+
+In `azure-data-tables`:
+```python
+from azure.data.tables import UpdateMode
+
+my_entity["Value"] += 5
+table_client.update_entity(my_entity)
+
+my_entity["StringProperty"] = "new_string"
+table_client.upsert_entity(my_entity, mode=UpdateMode.MERGE)
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/migration_guide.md:229]: I think we should have comparitive lines for each of the update scenarios - how to do merge vs replace has changed in T1 vs T2 and that is not illustrated here.

File: sdk/tables/azure-data-tables/migration_guide.md (hunk 22 above)
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/migration_guide.md:143]: There is no `query_tables` in T1, only list

Comment (annatisch) [reply sdk/tables/azure-data-tables/migration_guide.md:143]: The reason for the guide is not to show method signatures - rather it's to demonstrate "this is how I did X before, this is how I do it now".
So even if there's not an explicit method for it - in T1, one achieved this scenario by passing in a filter parameter, now one achieves it by calling a new method - so that's what our guide should show.

Comment (seankane-msft) [reply sdk/tables/azure-data-tables/migration_guide.md:143]: I don't think there is any ability to query the tables in T1. [Here](https://github.com/Azure/azure-cosmos-table-python/blob/master/azure-cosmosdb-table/azure/cosmosdb/table/tableservice.py#L434) is the method definition of `ServiceClient`. I don't see any way to add a filter to this call.

Comment (annatisch) [reply sdk/tables/azure-data-tables/migration_guide.md:140]: Ah. Haha - okay then, I will shut up now :)

**Review by mccoyp — APPROVED** [review]
I'm not very (or, at all) familiar with Tables, but the service-agnostic bits look great to me!

**Comment by seankane-msft** [comment]
https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/377/Updates-to-improve-track2-release-visibility

File: sdk/tables/azure-data-tables/migration_guide.md (hunk 6 above)
Comment (swathipil) [reply sdk/tables/azure-data-tables/migration_guide.md:77]: ^^ T1 vs. T2

File: sdk/tables/azure-data-tables/migration_guide.md (hunk 7 above)
Comment (swathipil) [reply sdk/tables/azure-data-tables/migration_guide.md:111]: yep, T1 vs. T2

**Review by swathipil — APPROVED** [review]
LGTM - I left a few comments previously about explaining differences b/w T1 vs. T2 (calling them cosmos vs. tables in case you weren't able to figure it out :P), but it looks like you addressed them

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 24]
```diff
@@ -0,0 +1,309 @@
...
+* [Important changes](#important-changes)
+    - [Package Names](#package-names-and-namespaces)
+    - [Client hierarchy and constructors](#client-hierarchy-and-constructors)
+    - [Authenticating Clients](#authenticating-clients)
+    - [Table Level Scenarios](#table-level-scenarios)
+    - [Entity Level Scenarios](#entity-level-scenarios)
+* [Additional Samples](#additional-samples)
+
+## Migration benefits
+
```
Comment (mario-guerra) [inline sdk/tables/azure-data-tables/migration_guide.md:19]: The information provided in this section isn't really describing benefits to migration. I would describe this information as more historical or informational because it provides the rationale for the change. What it *does not* provide is a solid case for why a developer should take the time to migrate existing code to the new package. They presumably already have code that works, why change it? 

I suggest retitling this portion to something like "History" or "Why a new package?" and following up with a section titled "migration benefits" or "Why upgrade?" that discusses the bennies to a developer of moving to this new package.

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 25]
```diff
@@ -0,0 +1,309 @@
...
+There were several areas of consistent feedback expressed across the Azure client library ecosystem. One of the most important is that the client libraries for different Azure services have not had a consistent approach to organization, naming, and API structure. Additionally, many developers have felt that the learning curve was difficult, and the APIs did not offer a good, approachable, and consistent onboarding story for those learning Azure or exploring a specific Azure service.
+
+To improve the development experience across Azure services, a set of uniform [design guidelines](https://azure.github.io/azure-sdk/general_introduction.html) were created for all languages to drive a consistent experience with established API patterns for all services. A set of [Python guidelines](https://azure.github.io/azure-sdk/python_design.html) were also introduced to ensure that Python clients have a natural and idiomatic feel with respect to the Python ecosystem. The new `azure-data-tables` follows these guidelines.
+
+### Cross Service SDK improvements
+
```
Comment (mario-guerra) [inline sdk/tables/azure-data-tables/migration_guide.md:27]: This section and the next section seem to describe actual benefits to a developer, i.e. new features, improved logging and authentication experience. Combine all these elements under a new heading titled 'Migration benefits" or "Why upgrade?"

File: sdk/tables/azure-data-tables/migration_guide.md [hunk 26]
```diff
@@ -0,0 +1,309 @@
+# Guide for migrating to `azure-data-tables` from `azure-cosmosdb-table`
+
+This guide is intended to assist in the migration to `azure-data-tables` from `azure-cosmosdb-table`. It will focus on side-by-side comparisons for similar operations between the two packages.
+
```
Comment (mario-guerra) [inline sdk/tables/azure-data-tables/migration_guide.md:4]: What about people using the old table storage SDK?

**Review by mario-guerra — COMMENTED** [review]
I've added a few comments regarding the discussion of benefits to migration, but the doc looks good overall.

File: sdk/tables/azure-data-tables/migration_guide.md (hunk 26 above)
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/migration_guide.md:4]: I'm not aware of an old table storage SDK, which one are you referring to?

Comment (mario-guerra) [reply sdk/tables/azure-data-tables/migration_guide.md:4]: Hmmm, good question. I know there's an existing table storage *service* and I was assuming that there was a corresponding SDK to go with it, but I haven't actually seen it. So if one doesn't actually exist then you can ignore my comment.

File: sdk/tables/azure-data-tables/migration_guide.md (hunk 25 above)
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/migration_guide.md:27]: The formatting of these sections is taken from the template [here](https://github.com/Azure/azure-sdk/blob/main/docs/policies/migration-guide-template.md)

Comment (mario-guerra) [reply sdk/tables/azure-data-tables/migration_guide.md:27]: I see. Who owns the template?

Comment (seankane-msft) [reply sdk/tables/azure-data-tables/migration_guide.md:27]: I believe @ramya-rao-a

Comment (mario-guerra) [reply sdk/tables/azure-data-tables/migration_guide.md:27]: I think the template could be tweaked but we have bigger fish to fry at the moment so I'll withdraw my comments for now. The doc LGTM
